src/pow.rs

```diff
@@ -36,55 +74,70 @@ impl Challenge {
     /// easier to get.
-    pub fn get(ip: &Ipv4Addr) -> Self {
-        let nonce = thread_rng().gen();
-        let expires_at = Instant::now() + TTL;
-        match nonce_set().cas(ip.to_bits(), None, Some((nonce, false))) {
+    pub fn get(ip: &Ipv4Addr, difficulty_if_not_present: u8) -> Self {
+        let challenge = Self {
+            nonce: thread_rng().gen(),
+            claimed: false,
+            expires_at: Instant::now() + SETTINGS.pow.challenge_duration,
+            difficulty: difficulty_if_not_present,
+        };
+        match challenge_set().cas(ip.to_bits(), None, Some(challenge.clone())) {
             Ok(None) => {
-                let nonce = Self { nonce, expires_at };
-                EvictionQueue::add_nonce(&nonce, *ip);
-                nonce
+                EvictionQueue::add_challenge(&challenge, *ip);
+                challenge
             }
+            Err(CasFailure {
+                actual: Some(challenge),
+                ..
+            }) => challenge,
             // Unreachable as this CAS will return a Some(..) only
             // in an Err.
             Ok(Some(_)) => unreachable!(),
-            Err(CasFailure { actual, .. }) => Self {
-                // safety: safe to unwrap actual as if it were None
-                // we would've got an Ok
-                nonce: actual.unwrap().0,
-                expires_at,
-            },
+            // Unreachable for same reason as above
+            Err(CasFailure { actual: None, .. }) => unreachable!(),
         }
     }
 
     /// Validates the proof of work solution by the client.
-    pub fn valid(
+    pub fn check_solution(
         ip: &Ipv4Addr,
-        difficulty: u8,
         solution: Solution,
     ) -> Result<(), OneOf<(NonceNotFound, BadProofOfWork, AlreadyClaimed)>> {
-        let ns = nonce_set();
+        let ns = challenge_set();
         let raw_ip = ip.to_bits();
-        let nonce = match ns.get(&raw_ip) {
-            Some((nonce, claimed)) if !claimed => nonce,
+        let mut challenge = match ns.get(&raw_ip) {
+            Some(nonce_data) if nonce_data.claimed == false => nonce_data,
             Some(_) => return err!(AlreadyClaimed),
             None => return err!(NonceNotFound),
         };
+
         let mut hasher = Sha256::new();
         hasher.update(b"strata faucet 2024");
-        hasher.update(nonce);
+        hasher.update(challenge.nonce);
         hasher.update(solution);
-        let pow_valid = count_leading_zeros(&hasher.finalize()) >= difficulty;
-        if pow_valid {
-            ns.insert(raw_ip, (nonce, true));
-            return Ok(());
+
+        // note, we mark the challenge as claimed here whether or not the
+        // proof of work is valid. This is because this effectively ratelimits
+        // the number of times a client can try to solve a challenge and waste
+        // our server resources.
+        challenge.claimed = true;
```

Comment:
Just to be clear.
`challenge.claimed` does not mean that the user has solved the PoW.
Just that is has "sent a challenge" and we need to validate and that's what happening in the lines below, right?